### ISSUE TITLE
OOM/infinite loop on malformed packets

### ISSUE THREAD

**Issue by JustinAzoff** [google/gopacket#439]
I ran gopacket through gofuzz for a while (forgot it was running)

```
// +build gofuzz
package gopacketfuzz

import (
	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
)

func Fuzz(data []byte) int {
	packet := gopacket.NewPacket(data, layers.LayerTypeEthernet, gopacket.Default)
	if packet == nil {
		return 0
	}
	layers := packet.Layers()
	if len(layers) == 0 {
		return 0
	}
	return 1
}
```

And found 3 packets that cause gopacket to use all memory before  being killed:

```
package main

import (
	"log"

	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
)

var packets = [][]byte{
	// workdir/crashers/f951a2ef03a3f465b0b75a4d9510a88d56404afa
	[]byte{
		0x5c, 0xfa, 0x00, 0x00, 0xfa, 0x6c, 0xbd, 0xe4, 0x19, 0x72, 0x1d, 0x1f,
		0x08, 0x00, 0x45, 0x00, 0x00, 0xad, 0x00, 0x00, 0x40, 0x00, 0x40, 0x59,
		0x6e, 0xef, 0x6b, 0x96, 0x61, 0x11, 0xb4, 0x96, 0xc0, 0xa8, 0x02, 0x04,
		0x00, 0xa2, 0x80, 0x00, 0x01, 0x1f, 0x08, 0x00, 0x45, 0x00, 0x6f, 0x76,
		0x65, 0x72, 0x66, 0x6c, 0x7f, 0xff, 0x12, 0x12, 0x00, 0x5a, 0xf6, 0x73,
		0xa2, 0xba, 0x81, 0x90, 0x00, 0x01, 0x40, 0x6f, 0x77, 0x20, 0x6f, 0x6e,
		0x20, 0x63, 0x68, 0x61, 0x72, 0x61, 0x63, 0x74, 0x00, 0x00, 0x00, 0x01,
		0x96, 0x20, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x20, 0x00, 0x6e, 0x00, 0x6f,
		0x77,
	},
	// workdir/crashers/b2fc2d04a6b15bd53f9d76f0a769038ea82d21b3
	[]byte{
		0x5c, 0xf9, 0x38, 0x95, 0x53, 0x6c, 0xbd, 0xe4, 0x19, 0x72, 0x1d, 0x1f,
		0x08, 0x00, 0x45, 0x00, 0x00, 0xad, 0x00, 0x00, 0x40, 0x00, 0x40, 0x59,
		0x6e, 0xef, 0x6b, 0x96, 0x61, 0x11, 0xb4, 0x96, 0xc0, 0xa8, 0x02, 0x04,
		0x00, 0xa2, 0x80, 0x00, 0x01, 0x1f, 0x08, 0x00, 0x45, 0x00, 0x6f, 0x00,
		0x45, 0x00, 0x5e, 0x65, 0x96, 0xc0, 0xa8, 0x02, 0x02, 0x97, 0x00, 0x35,
		0x10, 0x00, 0x00, 0x80, 0x00, 0x01, 0x1f, 0x08, 0x00, 0xa2, 0xba, 0x81,
		0x80, 0x00, 0x01, 0x1f, 0x08, 0x00, 0x45, 0x45, 0x00, 0x00, 0xad, 0x00,
		0x00, 0x40, 0x00, 0x40, 0x59, 0x6e, 0xef, 0x6b, 0x96, 0x61, 0x11, 0xb4,
		0x96, 0xc0, 0xa8, 0x02, 0x02, 0x97, 0x00, 0x35, 0x10, 0x00, 0x00, 0x80,
		0x00, 0x01, 0x1f, 0x08, 0x00, 0xa2, 0xba, 0x81, 0x80, 0x00, 0x01, 0x1f,
		0x08, 0x00, 0x45, 0x00, 0x6f, 0x76, 0x65, 0x72, 0x66, 0x6c, 0x6f, 0x77,
		0x20, 0x6f, 0x6e, 0x20, 0x63, 0x68, 0x61, 0x72, 0x61, 0x63, 0x74, 0x65,
		0x72, 0x11, 0xb4, 0x96, 0x20, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x20, 0x00,
		0x6e, 0x00, 0x00, 0x40, 0x00, 0x40, 0x61, 0x11, 0xb4, 0x96, 0xc0, 0xa8,
		0x02, 0x01, 0xc0, 0xa8, 0x02, 0x97, 0x00, 0x35, 0x10, 0x00, 0x00, 0x00,
		0xa2, 0xba, 0x81, 0x80, 0x00, 0x01, 0x1f, 0x08, 0x00, 0x45, 0x00, 0x00,
		0x6e, 0x00, 0x00, 0x40, 0x00, 0x40, 0x61, 0x11, 0xb4, 0x96, 0xc0, 0xa8,
		0x02, 0x01, 0xc0, 0xa8, 0x02, 0x97, 0x2f, 0xcf, 0xbf, 0xbd, 0x95, 0x53,
		0x6c, 0xc8, 0xd7, 0x19, 0x72, 0x1d, 0x1f, 0x08, 0x00, 0x45, 0x00, 0x00,
		0x6e, 0x00, 0x00, 0x40, 0x00, 0x97, 0x2f, 0xef, 0xbf, 0xbd, 0x95, 0x53,
		0x6c, 0xc8, 0xd7, 0x19, 0x72, 0x00, 0x6f, 0x76, 0x65, 0x72, 0x66, 0x6c,
		0x6f, 0x77, 0x20, 0x6f, 0x6e, 0x20, 0x63, 0x68, 0x61, 0x72, 0x61, 0x63,
		0x74, 0x65, 0x72, 0x11, 0xb4, 0x96, 0x20, 0x76, 0x61, 0x6c, 0x75, 0x65,
		0x20, 0x00, 0x6e, 0x00, 0x00, 0x40, 0x00, 0x40, 0x61, 0x72, 0x66, 0x6c,
		0x6f, 0x77, 0x20, 0x6f, 0x6e, 0x20, 0x63, 0x68, 0x61, 0x72, 0x61, 0x63,
		0x74, 0x65, 0x72, 0x11, 0xb4, 0x96, 0x20, 0x76, 0x61, 0x6c, 0x75, 0x65,
		0x20, 0x00, 0x6e, 0x00, 0x00, 0x40,
	},
	// workdir/crashers/a7952972f2fca30ad4cf41e698de981f9d59a487
	[]byte{
		0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30,
		0x08, 0x00, 0x45, 0x30, 0x30, 0x30, 0x30, 0x30, 0x40, 0x00, 0x30, 0x59,
		0x6e, 0xef, 0x6b, 0x96, 0x61, 0x11, 0xb4, 0x96, 0xc0, 0xa8, 0x02, 0x04,
		0x00, 0xa2, 0x80, 0x00, 0x01, 0x1f, 0x08, 0x00, 0x45, 0x00, 0x6f, 0x76,
		0x65, 0x72, 0x66, 0x6c, 0x6f, 0x77, 0x20, 0x6f, 0x6e, 0x20, 0x63, 0x68,
		0x61, 0xef, 0xbf, 0xbd, 0x12, 0x01, 0x00, 0x72, 0x61, 0x63, 0x74, 0x65,
		0x72, 0x11, 0xb4, 0x96, 0x20, 0x76, 0x61, 0x6c, 0x00, 0x00, 0x01, 0x00,
		0x6e, 0x00,
	},
}

func main() {
	for _, data := range packets {
		packet := gopacket.NewPacket(data, layers.LayerTypeEthernet, gopacket.Default)
		if packet == nil {
			continue
		}
		layers := packet.Layers()
		log.Printf("%+q", layers)
	}
}


```

The first packet sends decode into an infinite loop and never finishes, this is the 2nd packet:

```$ free -m
              total        used        free      shared  buff/cache   available
Mem:          15751         317       14112         235        1321       14896
Swap:             0           0           0
$ go run main.go

fatal error: runtime: out of memory

runtime stack:
runtime.throw(0x56a3ab, 0x16)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/panic.go:616 +0x81
runtime.sysMap(0xc746aa0000, 0xd0bb0000, 0x68c700, 0xbbcb78)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/mem_linux.go:216 +0x20a
runtime.(*mheap).sysAlloc(0x68bf60, 0xd0bb0000, 0x4994ee)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/malloc.go:470 +0xd4
runtime.(*mheap).grow(0x68bf60, 0x685d6, 0x0)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/mheap.go:907 +0x60
runtime.(*mheap).allocSpanLocked(0x68bf60, 0x685d6, 0xbbcb88, 0x7f0ebc4a7de8)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/mheap.go:820 +0x301
runtime.(*mheap).alloc_m(0x68bf60, 0x685d6, 0xffffffffffff0100, 0x7f0ebc4a7e18)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/mheap.go:686 +0x118
runtime.(*mheap).alloc.func1()
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/mheap.go:753 +0x4d
runtime.(*mheap).alloc(0x68bf60, 0x685d6, 0x7f0ebc010100, 0x41301c)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/mheap.go:752 +0x8a
runtime.largeAlloc(0xd0bac000, 0x440001, 0x7f0ebd41f000)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/malloc.go:826 +0x94
runtime.mallocgc.func1()
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/malloc.go:721 +0x46
runtime.systemstack(0x0)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/asm_amd64.s:409 +0x79
runtime.mstart()
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/proc.go:1175

goroutine 1 [running]:
runtime.systemstack_switch()
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/asm_amd64.s:363 fp=0xc4200595b8 sp=0xc4200595b0 pc=0x44e030
runtime.mallocgc(0xd0bac000, 0x52d2e0, 0xc420025201, 0xc420024000)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/malloc.go:720 +0x8a2 fp=0xc420059658 sp=0xc4200595b8 pc=0x40f752
runtime.growslice(0x52d2e0, 0xc69fae0000, 0x42cb199, 0x42cb199, 0x42cb19a, 0xc4a5a53500, 0x0, 0x0)
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/slice.go:179 +0x14a fp=0xc4200596c0 sp=0xc420059658 pc=0x43b34a
github.com/google/gopacket/layers.getLSAsv2(0x1020102f673a2ba, 0xc4200660ae, 0x23, 0x23, 0xbaffff0201020102, 0x1c0c7ffff00009e, 0xff0800b0b9ffff04, 0xc7ffff0c01c0c7ff, 0x4500544d4c1001c0)
	/home/justin/gocode/src/github.com/google/gopacket/layers/ospf.go:273 +0x427 fp=0xc420059788 sp=0xc4200596c0 pc=0x4cf457
github.com/google/gopacket/layers.(*OSPFv2).DecodeFromBytes(0xc4200680c0, 0xc420066092, 0x3f, 0x3f, 0x7f0ebd3cdfd0, 0xc42011e140, 0x60, 0x53bfe0)
	/home/justin/gocode/src/github.com/google/gopacket/layers/ospf.go:510 +0xc97 fp=0xc420059918 sp=0xc420059788 pc=0x4d18a7
github.com/google/gopacket/layers.decodingLayerDecoder(0x587620, 0xc4200680c0, 0xc420066092, 0x3f, 0x3f, 0x587ca0, 0xc42011e140, 0xec, 0x0)
	/home/justin/gocode/src/github.com/google/gopacket/layers/base.go:39 +0x9e fp=0xc420059978 sp=0xc420059918 pc=0x4a47de
github.com/google/gopacket/layers.decodeOSPF(0xc420066092, 0x3f, 0x3f, 0x587ca0, 0xc42011e140, 0xc42011e140, 0x5)
	/home/justin/gocode/src/github.com/google/gopacket/layers/ospf.go:666 +0x15c fp=0xc4200599d0 sp=0xc420059978 pc=0x4d2d4c
github.com/google/gopacket.DecodeFunc.Decode(0x571da8, 0xc420066092, 0x3f, 0x3f, 0x587ca0, 0xc42011e140, 0x40f1a9, 0xc4200d46e0)
	/home/justin/gocode/src/github.com/google/gopacket/decode.go:87 +0x58 fp=0xc420059a18 sp=0xc4200599d0 pc=0x494a78
github.com/google/gopacket.LayerType.Decode(0x7b, 0xc420066092, 0x3f, 0x3f, 0x587ca0, 0xc42011e140, 0x7f0ebd41f000, 0x0)
	/home/justin/gocode/src/github.com/google/gopacket/layertype.go:95 +0x91 fp=0xc420059a80 sp=0xc420059a18 pc=0x495e21
github.com/google/gopacket.(*LayerType).Decode(0xc42001a478, 0xc420066092, 0x3f, 0x3f, 0x587ca0, 0xc42011e140, 0x0, 0xc42001a478)
	<autogenerated>:1 +0x72 fp=0xc420059ad0 sp=0xc420059a80 pc=0x49a1f2
github.com/google/gopacket.(*eagerPacket).NextDecoder(0xc42011e140, 0x584960, 0xc42001a478, 0xc42001a478, 0x7f0ebd3cdfd0)
	/home/justin/gocode/src/github.com/google/gopacket/packet.go:443 +0x97 fp=0xc420059b20 sp=0xc420059ad0 pc=0x498427
github.com/google/gopacket/layers.decodeIPv4(0xc42006607e, 0x53, 0x53, 0x587ca0, 0xc42011e140, 0xc420000180, 0x300000002)
	/home/justin/gocode/src/github.com/google/gopacket/layers/ip4.go:282 +0x19d fp=0xc420059b98 sp=0xc420059b20 pc=0x4c843d
github.com/google/gopacket.DecodeFunc.Decode(0x571d40, 0xc42006607e, 0x53, 0x53, 0x587ca0, 0xc42011e140, 0x68, 0x540940)
	/home/justin/gocode/src/github.com/google/gopacket/decode.go:87 +0x58 fp=0xc420059be0 sp=0xc420059b98 pc=0x494a78
github.com/google/gopacket/layers.EthernetType.Decode(0x800, 0xc42006607e, 0x53, 0x53, 0x587ca0, 0xc42011e140, 0x7f0ebd41f000, 0x0)
	/home/justin/gocode/src/github.com/google/gopacket/layers/enums_generated.go:68 +0x79 fp=0xc420059c30 sp=0xc420059be0 pc=0x4bf409
github.com/google/gopacket/layers.(*EthernetType).Decode(0xc42001a470, 0xc42006607e, 0x53, 0x53, 0x587ca0, 0xc42011e140, 0x0, 0xc42001a470)
	<autogenerated>:1 +0x72 fp=0xc420059c80 sp=0xc420059c30 pc=0x4f1f02
github.com/google/gopacket.(*eagerPacket).NextDecoder(0xc42011e140, 0x5849c0, 0xc42001a470, 0xc42001a470, 0x7f0ebd3cdfd0)
	/home/justin/gocode/src/github.com/google/gopacket/packet.go:443 +0x97 fp=0xc420059cd0 sp=0xc420059c80 pc=0x498427
github.com/google/gopacket/layers.decodeEthernet(0xc420066070, 0x61, 0x61, 0x587ca0, 0xc42011e140, 0x3, 0xc420024070)
	/home/justin/gocode/src/github.com/google/gopacket/layers/ethernet.go:121 +0x178 fp=0xc420059d30 sp=0xc420059cd0 pc=0x4c1a98
github.com/google/gopacket.DecodeFunc.Decode(0x571cc0, 0xc420066070, 0x61, 0x61, 0x587ca0, 0xc42011e140, 0xc420000180, 0xc420064080)
	/home/justin/gocode/src/github.com/google/gopacket/decode.go:87 +0x58 fp=0xc420059d78 sp=0xc420059d30 pc=0x494a78
github.com/google/gopacket.LayerType.Decode(0x11, 0xc420066070, 0x61, 0x61, 0x587ca0, 0xc42011e140, 0x7f0ebd41f000, 0xc420059e38)
	/home/justin/gocode/src/github.com/google/gopacket/layertype.go:95 +0x91 fp=0xc420059de0 sp=0xc420059d78 pc=0x495e21
github.com/google/gopacket.(*LayerType).Decode(0xc42001a458, 0xc420066070, 0x61, 0x61, 0x587ca0, 0xc42011e140, 0x671401, 0xc42011e140)
	<autogenerated>:1 +0x72 fp=0xc420059e30 sp=0xc420059de0 pc=0x49a1f2
github.com/google/gopacket.(*eagerPacket).initialDecode(0xc42011e140, 0x584960, 0xc42001a458)
	/home/justin/gocode/src/github.com/google/gopacket/packet.go:447 +0x91 fp=0xc420059e80 sp=0xc420059e30 pc=0x498561
github.com/google/gopacket.NewPacket(0xc420066070, 0x61, 0x61, 0x584960, 0xc42001a458, 0xc400000000, 0x587ea0, 0xc42011e000)
	/home/justin/gocode/src/github.com/google/gopacket/packet.go:676 +0x24e fp=0xc420059ee0 sp=0xc420059e80 pc=0x4994ee
main.main()
	/home/justin/Sync/gocrap/gopacket_fuzz/boom/main.go:21 +0x1a0 fp=0xc420059f88 sp=0xc420059ee0 pc=0x4f8970
runtime.main()
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/proc.go:198 +0x212 fp=0xc420059fe0 sp=0xc420059f88 pc=0x429642
runtime.goexit()
	/home/justin/.gimme/versions/go1.10.1.linux.amd64/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc420059fe8 sp=0xc420059fe0 pc=0x450671
exit status 2
```

I can probably provide more info if needed, just wanted to get this filed.  Not sure if this would be better split out as 3 test cases inside ./layers/ospf_test.go or some place else.

**Comment by gconnell** [comment]
Awesome work, and thanks!  I'll try to work on these today.

**Comment by rstrand** [comment]
This should be fixed now #468 is merged

**Comment by JustinAzoff** [comment]
Cool.  I was thinking it would be neat to integrate gopacket with https://github.com/google/oss-fuzz, but I'm not sure if they support go projects.
